// apis/operatePendant.api.js
exports.operatePendant = function(req, res){
  var api = 'operatePendant'
    , sessID = req.headers.auth
    , operand // // valid values: 'create', 'update', 'remove', 'get', 'getby'
    , userID = Number(req.body.userID) || 0
    , body = {};
  if ( sessID === undefined ||
       sessID !== req.body.sessID ||
       userID === 0 ) {return res.end()};
  res.set('Content-Type', 'application/json');
  res.set('Auth', sessID);
  body.sessID = sessID;
  req.query.operand ? operand = req.query.operand : operand = req.body.operand;
//------------------------------------------------------

/* // Pendant description
pendants = {
  pndID: ,
  userID: , // Creator 
  loanAppID: ,
  loanUserID: ,
  clctAppID: ,
  clctUserID: ,
  amount: ,
  type: , // 'loan' or 'collect'
  status: , // 'offered', 'approved' or 'declined'
  overdueDate: date + pndActivePeriod,
  creationDate: date
}
*/

  var d = new Date()
    , pndActivePeriod = require('config').pendantProperties.activePeriod
    , pndID = req.body.pndID || [] // for 'approve', 'decline', 'cancel' and 'remove' operans
    , pendant = req.body.pendant || [] // for 'create' operands
    , filter = req.body.filter || {} // filter for 'get' and 'getby' operands
    , query = req.body.query || {} // select (find) query for getby operand
  ;

  if(!!!Array.isArray(pendant)){pendant = [].concat(pendant)};
  if(!!!Array.isArray(pndID)){pndID = [].concat(pndID)};

  session.get(sessID,{userID:1, validTill:1}, function(e,r){ // session validation
    if(e){return res.status(500).json({error: "API:"+api+":session validation: "+e})};
    if(r === null || r.userID !== userID){ // no sessID or different userID
      return errResponse(reason.invalidSessid, body, res);
    } else if(r.validTill < d.getTime()) { // sessID is expired
      return errResponse(reason.expiredSessid, body, res);
    } else { // sessID is OK
      switch(operand){
/*** create ***
* 1. Array of Objects {loanAppID, clctAppID, amount, type}.
* 2. Type is 'loan' or 'collect'
* 3. If type is 'loan' - update the loanAppID
* 4. clctAppID is always updated.
* 5. Returns Array of pendants.
***/
        case 'create':
          var pndCounter = pendant.length;
          if(!!!pndCounter){return res.end()};
          body.pendant = [];
          pendant.forEach(function(pnd){
            if(!!!pnd.amount     ||
               !!!pnd.loanAppID  ||
               !!!pnd.loanUserID ||
               !!!pnd.clctAppID  ||
               !!!pnd.clctUserID ||
               !!!pnd.type){return res.end()};
            pnd.status = 'offered';
            pnd.userID = userID;
            pnd.creationDate = d.getTime();
            pnd.overdueDate = d.setDate(d.getDate() + pndActivePeriod);
            count.get('pndID', function(e,r){
              if(e){return res.status(500).json({error: "API:"+api+": "+e})};
              pnd.pndID = r.value.seq;
              pendants.insert(pnd, function(e,r){
                if(e){return res.status(500).json({error: "API:"+api+": "+e})};
                delete pnd._id;
                body.pendant.push(pnd);
                if(pnd.type === 'loan'){
                  applications.get(pnd.loanAppID, {userID:1, pending:1, balance:1, pendants:1,}, function(e,a){
                    if(e){return res.status(500).json({error: "API:"+api+": "+e})};
                    if(!!!a){return res.end()};
                    a.pending += Number(pnd.amount);
                    a.balance -= Number(pnd.amount);                    
                    a.pendants.push(pnd.pndID);
                    applications.update(pnd.loanAppID, a, function(e){
                      if(e){return res.status(500).json({error: "API:"+api+":applications.update: "+e})};                        
                    });
                  }); //loanAppID
                } // pnd.type === 'loan'
                applications.get(pnd.clctAppID, {userID:1, pendants:1}, function(e,a){
                  if(e){return res.status(500).json({error: "API:"+api+": "+e})};
                  if(!!!a){return res.end()};
                  a.pendants.push(pnd.pndID);
                  pnd.clctUserID = a.userID;
                  applications.update(pnd.clctAppID, a, function(e){
                    if(e){return res.status(500).json({error: "API:"+api+":applications.update: "+e})};                        
                  });
                }); //clctAppID
                if(--pndCounter === 0){
                  body.result = body.pendant.length > 0 ? true : false;
                  body.result = true;
                  res.json(body);
                }
              }); //pendants.insert
            }); //count.get
          }); //pendant.forEach
        break; //create

/*** cancel ***
* 1. The user can cancel only it's own pendants
* 2. If status is 'declined' - just remove the pendant
* 3. If type is 'loan' or status is 'approved' - update the loan application
* 4. Always update the collect application.
***/
        case 'cancel':
          var pndCounter = pndID.length;
          if(!!!pndCounter){return res.end()};
          body.pndID = [];
          pndID.forEach(function(pnd){
            pendants.get(pnd, {}, function(e,p){
              if(e){return res.status(500).json({error: "API:"+api+":pendants.get: "+e})};
              if(!!p && p.userID === userID){
                body.pndID.push(pnd);
                if(p.status !== 'declined'){
                  if(p.type === 'loan' || p.status === 'approved'){
                    applications.get(p.loanAppID, {pending:1, balance:1, pendants:1}, function(e,a){
                      if(e){return res.status(500).json({error: "API:"+api+":applications.get: "+e})};
                      if(!!!a){return res.end()};
                      a.pendants.splice(a.pendants.indexOf(pnd),1);
                      a.pending -= Number(p.amount);
                      a.balance += Number(p.amount);
                      applications.update(p.loanAppID, a, function(e){
                        if(e){return res.status(500).json({error: "API:"+api+":applications.update: "+e})};                        
                      });
                    });                    
                  } // p.type === 'loan' || p.status === 'approved'. 
                  applications.get(p.clctAppID, {pendants:1}, function(e,a){
                    if(e){return res.status(500).json({error: "API:"+api+":applications.get: "+e})};
                    if(!!!a){return res.end()};
                    a.pendants.splice(a.pendants.indexOf(pnd),1);
                    applications.update(p.clctAppID, a, function(e){
                      if(e){return res.status(500).json({error: "API:"+api+":applications.update: "+e})};                        
                    });
                  });                  
                } // p.status !== 'declined'. When 'declined' - this has already done.
                pendants.remove(pnd, function(e){
                  if(e){return res.status(500).json({error: "API:"+api+":pendants.remove: "+e})};
                  // send message to the appropriate userID
                }); // pendants.remove
              }; // !!p && p.userID === userID
              if(--pndCounter === 0){
                body.result = body.pndID.length > 0 ? true : false;
                res.json(body);
              }
            }); // pendants.get
          }); // pndID.forEach
        break; // cancel

/*** approve ***
* 1. Process only 'offered' status.
* 2. If the type is 'loan' just set a status as 'approved'
* 3. If the type is 'collect' - update the loan application and set a status.
***/
        case 'approve':
          var pndCounter = pndID.length;
          if(!!!pndCounter){return res.end()}; // must be > 0
          body.pendant = [];
          body.pndID = [];
          pndID.forEach(function(pnd){
            pendants.get(pnd, {}, function(e,p){
              if(e){return res.status(500).json({error: "API:"+api+":pendants.get: "+e})};
              if(!!p && p.status === 'offered' &&
                 userID === (p.loanUserID === p.userID ? p.clctUserID : p.loanUserID) // User can't approve its own pendant.
                ){
                p.status = 'approved';
                body.pendant.push(p);
                body.pndID.push(pnd);                
                pendants.updateStatus(pnd, 'approved', function(e,r){
                  if(e){return res.status(500).json({error: "API:"+api+":pendants.updateStatus: "+e})};
                  // send message to the appropriate userID
                });
                if(p.type === 'collect'){
                  applications.get(p.loanAppID, {pending:1, balance:1, pendants:1,}, function(e,a){
                    if(e){return res.status(500).json({error: "API:"+api+":applications.get: "+e})};
                    if(!!!a){return res.end()};
                    a.pending += Number(p.amount);
                    a.balance -= Number(p.amount);                    
                    a.pendants.push(pnd);
                    applications.update(p.loanAppID, a, function(e){
                      if(e){return res.status(500).json({error: "API:"+api+":applications.update: "+e})};                        
                    });
                  }); //loanAppID
                } // p.type === 'collect'
              }; // p.status === 'offered'
              if(--pndCounter === 0){
                body.result = body.pndID.length > 0 ? true : false;
                res.json(body);
              };
            }); // pendants.get
          }); // pndID.forEach
        break; // approve

/*** decline ***
* 1. It is possible to decline already approved application.
* 2. The pendant is removed from the application's list. From loan and from collect applications
* 3. The status is set as 'declined'
* 4. User can remove it's declined pendants.
***/
        case 'decline':
          var pndCounter = pndID.length;
          if(!!!pndCounter){return res.end()};
          body.pendant = [];
          body.pndID = [];
          pndID.forEach(function(pnd){
            pendants.get(pnd, {}, function(e,p){
              if(e){return res.status(500).json({error: "API:"+api+":pendants.get: "+e})};
              if(!!p && p.status === 'offered' && // It is possible to decline only offered pendant.
                 userID === (p.loanUserID === p.userID ? p.clctUserID : p.loanUserID) // User can't decline its own pendant.
                ){ 
                if(p.type === 'loan'){                  
                  applications.get(p.loanAppID, {pending:1, balance:1, pendants:1}, function(e,a){
                    if(e){return res.status(500).json({error: "API:"+api+":applications.get: "+e})};
                    if(!!!a){return res.end()};
                    a.pendants.splice(a.pendants.indexOf(pnd),1);
                    a.pending -= Number(p.amount);
                    a.balance += Number(p.amount);
                    applications.update(p.loanAppID, a, function(e){
                      if(e){return res.status(500).json({error: "API:"+api+":applications.update: "+e})};                        
                    });
                  });
                } // p.type === 'loan'
                applications.get(p.clctAppID, {pendants:1}, function(e,a){
                  if(e){return res.status(500).json({error: "API:"+api+":applications.get: "+e})};
                  if(!!!a){return res.end()};
                  a.pendants.splice(a.pendants.indexOf(pnd),1);
                  applications.update(p.clctAppID, a, function(e){
                    if(e){return res.status(500).json({error: "API:"+api+":applications.update: "+e})};                        
                  });
                });
                pendants.updateStatus(pnd, 'declined', function(e,r){
                  if(e){return res.status(500).json({error: "API:"+api+":pendants.updateStatus: "+e})};
                  // send message to appropriate userID.
                });                             
                p.status = 'declined';
                body.pendant.push(p);
                body.pndID.push(pnd);                
              } // p.status !== 'declined' 
              if(--pndCounter === 0){
                body.result = body.pndID.length > 0 ? true : false;
                res.json(body);
              };
            }); // pendants.get
          }); // pndID.forEach
        break; // decline

/*** remove ***
* 1. User can remove only it's declined pendants
***/
        case 'remove': // Only for declined pendants
          var pndCounter = pndID.length;
          if(!!!pndCounter){return res.end()};
          body.pndID = [];
          pndID.forEach(function(pnd){
            pendants.get(pnd, {status:1, userID:1}, function(e,p){
              if(e){return res.status(500).json({error: "API:"+api+":pendants.get: "+e})};
              if(!!p && p.status === 'declined' && p.userID === userID){
                body.pndID.push(pnd);                
                pendants.remove(pnd, function(e){
                  if(e){return res.status(500).json({error: "API:"+api+":pendants.remove: "+e})};
                }); //pendants.remove
              }
              if(--pndCounter === 0){
                body.result = body.pndID.length > 0 ? true : false;
                res.json(body);
              };                
            }); // pendants.get
          }); // pndID.forEach
        break; // remove

/*** getUserPendants and getAssignedPendants ***
*  Return list of user related pendants.
***/
        case 'getUserPendants':
          pendants.getUserPendants(userID, {}, function(e,r){
            if(e){return res.status(500).json({error: "API:"+api+":pendants.getUserPendants: "+e})};
            body.pendant = r;
            body.result = r.length > 0 ? true: false;
            res.json(body);
          });
        break;
        case 'getAssignedPendants':
          pendants.getAssignedPendants(userID, {}, function(e,r){
            if(e){return res.status(500).json({error: "API:"+api+":pendants.getAssignedPendants: "+e})};
            body.pendant = r;
            body.result = r.length > 0 ? true : false;
            res.json(body);
          });
        break;        
/*** get and getby ***
* General APIs for DB queries
***/
        case 'get':
          if(pndID.length !== 1){return res.end()};
          pndID = pndID.shift(); // Extract pndID from the List
          pendants.get(pndID, filter, function(e,r){
            if(e){return res.status(500).json({error: "API:"+api+":get: "+e})};
            body.result = r ? true : false;
            if(r){
              body.result = true;
              body.pndID = pndID;
              body.pendant = r;
            } else {
              body.result = false;
              body.reason = reason.notFoundInDB;              
            };
            res.json(body);
          });
        break;
        case 'getby':
          pendants.getBy(query, filter, function(e,r){
            if(e){return res.status(500).json({error: "API:"+api+":getby: "+e})};
            if(r.length){
              body.result = true;
              body.pndID = r.map(function(el){return el.pndID});
              body.pendant = r;
            } else {
              body.result = false;
              body.reason = reason.notFoundInDB;              
            }
            res.json(body);
          });
        break;        
        default:
          res.status(500).json({error: "API:"+api+": Operand "+operand+" is not supported"});
      } // switch
    } // if-else
  }); // session validation
}
