// models/apparchive.js
// Applications archive model - APIs for the apparchive collection in the wwmmorg DB
var collectionName = 'apparchive',
    model = function(){

  this.db = {};
  this.collection = {};

  this.setDB = function(db) {
    this.db = db;
    this.collection = db.collection(collectionName);
    return this;
  };

  this.insert = function(data, callback) {
    return this.collection.insertOne(data, {w:1,j:true}, callback || function(){});
  };

  this.get = function(appID, filter, callback) { // By appID
    var appID = Number(appID);
    filter ? filter._id = 0 : filter = {_id: 0};
    return this.collection.find({appID: appID}, filter)
               .limit(1)
               .next(callback || function(){});
  };
  
  this.getBy = function(query, filter, callback) {
    var query = query || {};
    filter ? filter._id = 0 : filter = {_id: 0};
    return this.collection.find(query, filter)
               .toArray(callback || function(){});
  };
  
  this.getUserApps = function(userID, filter, callback) { // List of applications by userID
    var userID = Number(userID);
    filter ? filter._id = 0 : filter = {_id: 0};
    return this.collection.find({userID: userID}, filter).toArray(callback);
  };

  this.remove = function(appID, callback) { // Only for testing purpose
    if(!!!Array.isArray(appID)){appID = [appID]};
    appID = appID.map(function(el){return Number(el)});
    return this.collection.deleteMany({appID: {$in: appID}}, {w:1,j:true}, callback || function(){});
  };
};

module.exports = new model();
