// models/enrolls.js
// Enrolls model - APIs for the enrolls collection in the wwmmorg DB
var collectionName = 'enrolls'
  , model = function(){

  this.db = {};
  this.collection = {};

  this.setDB = function(db) {
    this.db = db;
    this.collection = db.collection(collectionName);
    return this;
  };

  this.insert = function(data, callback) {
    var num = Number(data.number) || 0
      , str = data.string || ''
      , email = data.email || ''
      , time = data.enrolmentDate || new Date().getTime()
      ;

      return this.collection.findAndModify({email: email},
                                           [["email",1]],
                                           {string: str, number: num, email: email, enrolmentDate: time}, 
                                           {w:1,j:true,upsert:true,new:true},
                                           callback || function(){}
                                          );
  };

  this.get = function(data, callback) { // By string, number and e-mail
    var num = Number(data.number) || 0
      , str = data.string || ''
      , email = data.email || ''
      ;
//    return this.collection.find({string: str, number: num, email: email},
    return this.collection.find({email: email, number: num, string: str},
                                {_id:0, enrolmentDate:0})
//               .limit(1)
               .next(callback || function(){});
  };

  this.remove = function(email, callback) { // remove all by e-mail
    return this.collection.deleteMany({email: email}, {w:1,j:true}, callback || function(){});
  };

  this.clean = function(callback) { // remove all by e-mail
    var midnight = new Date(new Date().toDateString()).getTime();
    return this.collection.deleteMany({enrolmentDate: {$lt: midnight}}, {w:1,j:true}, callback || function(){});
  };

// Only for testing purposes:
  this.bring = function(email, callback) { // By string, number and e-mail
    return this.collection.find({email: email},
                                {_id:0, enrolmentDate:0})
               .limit(1)
               .next(callback || function(){});
  };
}

module.exports = new model();
