// models/pendants.js
// Template model - APIs for the template collection in the wwmmorg DB
var collectionName = 'pendants',
    model = function(){

  this.db = {};
  this.collection = {};

  this.setDB = function(db) {
    this.db = db;
    this.collection = db.collection(collectionName);
    return this;
  };

  this.insert = function(data, callback) {
    return this.collection.insertOne(data, {w:1,j:true}, callback || function(){});
  };

  this.get = function(pndID, filter, callback) {
    var pndID = Number(pndID);
    filter ? filter._id = 0 : filter = {_id: 0};     
    return this.collection.find({pndID: pndID}, filter)
               .limit(1)
               .next(callback || function(){});
  };
  
  this.getBy = function(query, filter, callback) {
    var query = query || {};
    filter ? filter._id = 0 : filter = {_id: 0};
    return this.collection.find(query, filter)
               .toArray(callback || function(){});
  };

  this.remove = function(pndID, callback) {
    var pndID = Number(pndID);
    return this.collection.deleteOne({pndID: pndID}, {w:1,j:true}, callback || function(){});
  };
  
  this.removeDeclined = function(userID, pendants, callback) {
    var userID = Number(userID)
      , pendants = pendants.map(function(id){return Number(id)})
    ;
    return this.collection.deleteMany({userID: userID, status: 'declined', pndID: {$in: pendants}}, {w:1,j:true}, callback || function(){});
  };

  this.updateStatus = function(pndID, Status, callback) {
    var pndID = Number(pndID);
    return this.collection.updateOne({pndID: pndID}, {$set: {status: Status}}, {w:1,j:true}, callback || function(){});
  };

  this.getUserPendants = function(userID, filter, callback){
    var userID = Number(userID);  
    filter ? filter._id = 0 : filter = {_id: 0};
    return this.collection.find({userID: userID}, filter)
               .toArray(callback || function(){});
  };

  this.getAssignedPendants = function(userID, filter, callback){
    var userID = Number(userID);  
    filter ? filter._id = 0 : filter = {_id: 0};
    return this.collection.find({$and: [{$or: [{loanUserID: userID}, {clctUserID: userID}]}, {userID: {$ne: userID}}]}, filter)
               .toArray(callback || function(){});
    
  };
  
  this.getOverduePendants = function(callback){
    var now = new Date().getTime();
    return this.collection.find({$and: [{status: 'offered'},
                                        {overdueDate: {$lt: now}}
                                ]},
                                {_id: 0})
               .sort({overdueDate: 1})
               .toArray(callback);
  };
};

module.exports = new model();
