// models/transactions.js
// Transactions model - APIs for the transactions collection in the wwmmorg DB
var collectionName = 'transactions',
    model = function(){

  this.db = {};
  this.collection = {};

  this.setDB = function(db) {
    this.db = db;
    this.collection = db.collection(collectionName);
    return this;
  };

  this.insert = function(data, callback) {
    return this.collection.insertOne(data, {w:1,j:true}, callback || function(){});
  };

  this.update = function(transID, data, callback) {
    var transID = Number(transID);
    if(Object.keys(data).length === 0){data.transID = transID};
    return this.collection.updateOne({transID: transID}, {$set: data}, {w:1,j:true}, callback || function(){});
  };
  
  this.get = function(transID, filter, callback) { // By transID
    var transID = Number(transID);
    filter ? filter._id = 0 : filter = {_id: 0};
    return this.collection.find({transID: transID}, filter)
               .limit(1)
               .next(callback || function(){});
  };
  
  this.remove = function(transID, callback) { // Only for testing purpose
    if(!!!Array.isArray(transID)){transID = [transID]};
    transID = transID.map(function(el){return Number(el)});
    return this.collection.deleteOne({transID: {$in: transID}}, {w:1,j:true}, callback || function(){});
  };

  this.updateStatus = function(transID, Status, callback) {
    var transID = Number(transID);
    return this.collection.updateOne({transID: transID}, {$set: {status: Status}}, {w:1,j:true}, callback || function(){});
  };

  this.getUserTransactions = function(userID, filter, callback) { // List of transactions by userID
    var userID = Number(userID);
    filter ? filter._id = 0 : filter = {_id: 0};
    return this.collection.find({$or: [{loanUserID: userID}, {clctUserID: userID}]}, filter).toArray(callback);
  };
};

module.exports = new model();
